RHDEVDOCS-4064 - Loki Integration (#47703)
openshift/openshift-docs: modules/cluster-logging-loki-deploy.adoc

.. Verify that you installed the LokiOperator. Visit the *Operators* → *Installed Operators* page and look for "LokiOperator."

.. Ensure that *LokiOperator* is listed with *Status* as *Succeeded* in all the projects.
+
. Create a Secret YAML file that uses the `access_key_id` and `access_key_secret` fields to specify your base64-encoded AWS credentials. For example:
+
[source,yaml]
----
apiVersion: v1
kind: Secret
metadata:
  name: logging-loki-s3
  namespace: openshift-logging
data:
  access_key_id: QUtJQUlPU0ZPRE5ON0VYQU1QTEUK
  access_key_secret: d0phbHJYVXRuRkVNSS9LN01ERU5HL2JQeFJmaUNZRVhBTVBMRUtFWQo=
  bucketnames: czMtYnVja2V0LW5hbWU= <1>
  endpoint: aHR0cHM6Ly9zMy5ldS1jZW50cmFsLTEuYW1hem9uYXdzLmNvbQ== <2>
----
<1> Base64-encoded bucket name
<2> Base64-encoded S3 API endpoint
+
. Create `LokiStack` Custom Resource
+
[source,yaml]
----
  apiVersion: loki.grafana.com/v1
  kind: LokiStack
  metadata:
    name: logging-loki
    namespace: openshift-logging
  spec:
    size: 1x.small
    storage:
      schemas:
      - version: v12
        effectiveDate: 2022-06-01
      secret:
        name: logging-loki-s3
        type: s3
    storageClassName: gp2
    tenants:

>       mode: openshift-logging

Reviewer: Can we add a note regarding storage schema version. If the user has not added any storage schema in the CR, then storage schema is created with v11.
To upgrade to v12, we need to add another new section in the `schemas` list.

For e.g. 
```
    storage:
      schemas:
      - version: v11
        effectiveDate: 2020-06-01
      - version: v12
      	effectiveDate: <future date>
```

Author: I am open to adding it later, but we don't mention storage schema at all anywhere else in the logging docs, is there somewhere else in the OpenShift Docs that we might be able to link to or pull from to contextualize that option for users?

Reviewer: Sounds good! this schema version is referenced to Loki only.
Reference: https://grafana.com/docs/loki/latest/operations/storage/schema/

Reviewer 2: Maybe the Grafana docs?
https://grafana.com/docs/loki/latest/operations/storage/schema/#changing-the-schema

Author: Added to additional resources.

----
+
.. Apply the configuration
+
[source,terminal]
----
oc apply -f logging-loki.yaml
----
+

> . Create or edit a `ClusterLogging` Custom Resource

Reviewer 3: ```suggestion
. Create or edit a `ClusterLogging` CR:
```

+
[source,yaml]
----
  apiVersion: logging.openshift.io/v1
  kind: ClusterLogging
  metadata:
    name: cr-lokistack
    namespace: openshift-logging
  spec:
   logStore:
    type: lokistack
    lokistack:

>     name: logging-loki

Reviewer: name needs to be under lokistack.
```
  logStore:
    lokistack:
      name: <lokistack-name>
```

   collection:
    type: vector
----
+
.. Apply the configuration
+
[source,terminal]
----
oc apply -f cr-lokistack.yaml
----
+

> . Enable the RedHat OpenShift Logging Console Plugin

Reviewer 3: ```suggestion
. Enable the RedHat OpenShift Logging Console Plugin:
```

.. In the {product-title} web console, click *Operators* -> *Installed Operators*
.. Select the *RedHat OpenShift Logging* Operator

> .. Under Console plugin, click *Disabled*

Reviewer 3: Periods.

.. Select *Enable* and then *Save*. This change will restart the 'openshift-console' pods.
.. After the pods restart, you will recieve a notification that a web console update is available, prompting you to refresh.

> .. After refreshing the web console, click *Observe* from the left main menu. A new option for *Logs* will be available to you.

Reviewer 4: We just had a meeting on the topic of the new logging UI. Would it make sense to link to the [LogQL documentation](https://grafana.com/docs/loki/latest/logql/) from this document somewhere?

Author: Added to additional resources.

.. Select an *Approval Strategy*.
+
* The *Automatic* strategy allows Operator Lifecycle Manager (OLM) to automatically update the Operator when a new version is available.
+
* The *Manual* strategy requires a user with appropriate credentials to approve the Operator update.

.. Click *Install*.

.. Verify that you installed the LokiOperator. Visit the *Operators* → *Installed Operators* page and look for *LokiOperator*.

.. Ensure that *LokiOperator* is listed with *Status* as *Succeeded* in all the projects.
+
. Create a `Secret` YAML file that uses the `access_key_id` and `access_key_secret` fields to specify your base64-encoded AWS credentials. For example:
+
[source,yaml]
----
apiVersion: v1
kind: Secret
metadata:
  name: logging-loki-s3
  namespace: openshift-logging
data:
  access_key_id: QUtJQUlPU0ZPRE5ON0VYQU1QTEUK
  access_key_secret: d0phbHJYVXRuRkVNSS9LN01ERU5HL2JQeFJmaUNZRVhBTVBMRUtFWQo=
  bucketnames: czMtYnVja2V0LW5hbWU= <1>
  endpoint: aHR0cHM6Ly9zMy5ldS1jZW50cmFsLTEuYW1hem9uYXdzLmNvbQ== <2>
----
<1> Base64-encoded bucket name
<2> Base64-encoded S3 API endpoint
+

> . Create `LokiStack` custom resource:

Reviewer 3: ```suggestion
. Create the `LokiStack` custom resource:
```

+
[source,yaml]
----
  apiVersion: loki.grafana.com/v1
  kind: LokiStack
  metadata:
    name: logging-loki
    namespace: openshift-logging
  spec:
    size: 1x.small
    storage:
      schemas:
      - version: v12
        effectiveDate: 2022-06-01
      secret:
        name: logging-loki-s3
        type: s3
    storageClassName: gp2
    tenants:
      mode: openshift-logging
----
+
.. Apply the configuration:
+
[source,terminal]
----
oc apply -f logging-loki.yaml
----
+
. Create or edit a  Create or edit a `ClusterLogging` CR:
+
[source,yaml]
----
  apiVersion: logging.openshift.io/v1
  kind: ClusterLogging
  metadata:
    name: cr-lokistack
    namespace: openshift-logging
  spec:
     logStore:
       lokistack:

>          name: <lokistack-name>

Reviewer: Can we mention the 'logging-loki' as the lokistack name since we have been using that a reference in earlier CR's.

Author: Corrected.

// Module is included in the following assemblies:
//cluster-logging-loki.adoc
:_content-type: PROCEDURE
[id="logging-loki-deploy_{context}"]
= Deploying the LokiStack
You can use the {product-title} web console to install the LokiOperator.

.Prerequisites

* {logging-title-uc} Operator 5.5

> * S3 bucket for log storage

Reviewer 4: Should we mention somewhere that other object storages are implemented (see [here](https://loki-operator.dev/docs/object_storage.md/)) but currently "not supported" (= untested)?

Reviewer 5: @periklis  @xperimental Will we support the other object storage beside of s3?   QE had verified the chunk be written into these storage.

Author: Documented = supported, generally speaking. So are they not supported as in we don't plan to support them, or considered "tech preview"?

Reviewer 4: As far as I am aware we planned to take S3 to GA first and the follow-up with the other object storages later on. The implementation turned out to be pretty simple and has already happened and I was not aware that other object storages aside from the S3 / S3-compatible (ODF, Minio, ...) storages have already been tested. Also we might be talking of different things, for GCS I know that they _also_ have a S3-compatible API, but that's not what we are talking about when talking about using GCS as an object storage, but rather the "GCS native" API. Similar offers _might_ be available from the other providers as well.

For now, I think we could go with documenting only S3 and maybe linking to the loki-operator documentation for "further information", but my information might also be outdated.

Reviewer 5: Got it. Thanks

.. Under *Installed Namespace*, select *openshift-operators-redhat*.
+
You must specify the `openshift-operators-redhat` namespace. The `openshift-operators`
namespace might contain Community Operators, which are untrusted and could publish
a metric with the same name as an {product-title} metric, which would cause
conflicts.

.. Select *Enable operator recommended cluster monitoring on this namespace*.
+
This option sets the `openshift.io/cluster-monitoring: "true"` label in the Namespace object.
You must select this option to ensure that cluster monitoring
scrapes the `openshift-operators-redhat` namespace.

.. Select an *Approval Strategy*.
+
* The *Automatic* strategy allows Operator Lifecycle Manager (OLM) to automatically update the Operator when a new version is available.
+
* The *Manual* strategy requires a user with appropriate credentials to approve the Operator update.

.. Click *Install*.

.. Verify that you installed the LokiOperator. Visit the *Operators* → *Installed Operators* page and look for "LokiOperator."

.. Ensure that *LokiOperator* is listed with *Status* as *Succeeded* in all the projects.
+

> . Create a Secret YAML file that uses the aws_access_key_id and aws_secret_access_key fields to specify your base64-encoded AWS credentials. For example:

Reviewer 4: ```suggestion
. Create a Secret YAML file that uses the access_key_id and access_key_secret fields to specify your base64-encoded AWS credentials. For example:
```

+
[source,yaml]
----
apiVersion: v1
kind: Secret
metadata:
  name: logging-loki-s3
  namespace: openshift-logging
data:

>   aws_access_key_id: QUtJQUlPU0ZPRE5ON0VYQU1QTEUK
>   aws_secret_access_key: d0phbHJYVXRuRkVNSS9LN01ERU5HL2JQeFJmaUNZRVhBTVBMRUtFWQo=

Reviewer 4: This configuration is not sufficient, the secret also needs to contain the bucket name, endpoint and optionally region:

```suggestion
  access_key_id: QUtJQUlPU0ZPRE5ON0VYQU1QTEUK
  access_key_secret: d0phbHJYVXRuRkVNSS9LN01ERU5HL2JQeFJmaUNZRVhBTVBMRUtFWQo=
  bucketnames: czMtYnVja2V0LW5hbWU= # base64-encoded bucket name
  endpoint: aHR0cHM6Ly9zMy5ldS1jZW50cmFsLTEuYW1hem9uYXdzLmNvbQ== # base64-encoded S3 API endpoint
```

>   aws_secret_access_key: d0phbHJYVXRuRkVNSS9LN01ERU5HL2JQeFJmaUNZRVhBTVBMRUtFWQo=

Author: @anpingli - @xperimental - I'm confused, are all of these storage configurations supported or not?

>   aws_secret_access_key: d0phbHJYVXRuRkVNSS9LN01ERU5HL2JQeFJmaUNZRVhBTVBMRUtFWQo=

Reviewer: @libander These storage configurations are supported. it's just on different platforms.

----
+
. Create `LokiStack` Custom Resource
+
[source,yaml]
----
  apiVersion: loki.grafana.com/v1
  kind: LokiStack
  metadata:
    name: logging-loki
    namespace: openshift-logging
  spec:
    size: 1x.small
    storage:
      schemas:
      - version: v12
        effectiveDate: 2022-06-01
      secret:
        name: logging-loki-s3
        type: s3
    storageClassName: gp2
    tenants:
      mode: openshift-logging
    rules:
      enabled: true
      selector:
        matchLabels:
          openshift.io/cluster-monitoring: "true"
      namespaceSelector:
        matchLabels:

>           openshift.io/cluster-monitoring: "true"

Reviewer 4: Can you mark the ruler component as "optional"? It's only needed if someone wants to try out the log-based metrics / alerts (`RecordingRule` / `AlertingRule` CRDs, which are still tech-preview).

Another option would be to leave the ruler out of the basic configuration completely, if only GA features should make it to this howto document.

Author: Thank you for the feedback, I'm going to remove the rules section in lieu of notating that it's tech-preview only for the sake of clarity.

----
+
.. Apply the configuration
+
[source,terminal]
----
oc apply -f logging-loki.yaml
----
+
. Create or edit a `ClusterLogging` Custom Resource
+
[source,yaml]
----
  apiVersion: logging.openshift.io/v1
  kind: ClusterLogging
  metadata:
    name: cr-lokistack
    namespace: openshift-logging
  spec:
   logStore
    type: lokistack
    lokistack:
     name: logging-loki
   collection:

>     logs:
>      type: vector

Reviewer 4: `logs` is deprecated, the `type` can be moved up one level:

```suggestion
    type: vector
```

I also noticed that the indentation on the YAML is different between the `metadata` and `spec` fields (1 space vs 2 spaces). This _should_ work but could be confusing.

// Module is included in the following assemblies:
//cluster-logging-loki.adoc
:_content-type: PROCEDURE
[id="logging-loki-deploy_{context}"]
= Deploying the Lokistack
You can use the {product-title} web console to install the LokiOperator.

.Prerequisites

> * {logging-title-uc}: 5.4

Reviewer 4: Should this be "5.5"?

```suggestion
* {logging-title-uc}: 5.5
```

.. Under *Installation Mode*, select *All namespaces on the cluster*.

.. Under *Installed Namespace*, select *openshift-operators-redhat*.
+
You must specify the `openshift-operators-redhat` namespace. The `openshift-operators`
namespace might contain Community Operators, which are untrusted and could publish
a metric with the same name as an {product-title} metric, which would cause
conflicts.

.. Select *Enable operator recommended cluster monitoring on this namespace*.
+
This option sets the `openshift.io/cluster-monitoring: "true"` label in the Namespace object.
You must select this option to ensure that cluster monitoring
scrapes the `openshift-operators-redhat` namespace.

.. Select an *Approval Strategy*.
+
* The *Automatic* strategy allows Operator Lifecycle Manager (OLM) to automatically update the Operator when a new version is available.
+
* The *Manual* strategy requires a user with appropriate credentials to approve the Operator update.

.. Click *Install*.

.. Verify that you installed the LokiOperator. Visit the *Operators* → *Installed Operators* page and look for "LokiOperator."

> .. Ensure that *LokiOperator* is listed in all the projects whose *Status* is *Succeeded*.

Reviewer 4: I don't understand this sentence. For me it makes more sense this way:

```suggestion
.. Ensure that *LokiOperator* is listed with *Status* as *Succeeded* in all the projects.
```

> .. Ensure that *LokiOperator* is listed in all the projects whose *Status* is *Succeeded*.

Reviewer 5: Could we provide some lokistack example and list the supported t-shirt size and supported Object storage type here?
